components/ai_chat/core/browser/conversation_handler.cc

```diff
@@ -1720,31 +1720,81 @@ void ConversationHandler::OnContentTaskStarted(int32_t tab_id) {
-  } else if (tool_loop_was_paused_by_tool_provider_) {
-    // Only try to resume the tool loop once per un-pause.
-    tool_loop_was_paused_by_tool_provider_ = false;
-    // If we're currently still executing a tool and we're unpausing, we don't
-    // need to kick off a new request and can wait for the callback of the tool.
-    if (!is_tool_use_in_progress_) {
-      MaybeRespondToNextToolUseRequest();
+void ConversationHandler::PauseTask() {
+  // Allow pausing even if no task is active yet. This handles the case where
+  // a tool provider pauses during the generation callback, before
+  // MaybeRespondToNextToolUseRequest has initialized the task state.
+  // Don't allow moving from stopped to paused as that will create a state that
+  // can be resumed.
+  if (tool_use_task_state_ != mojom::TaskState::kPaused &&
+      tool_use_task_state_ != mojom::TaskState::kStopped) {
+    tool_use_task_state_ = mojom::TaskState::kPaused;
+    OnToolUseTaskStateChanged();
+    for (auto& tool_provider : tool_providers_) {
+      tool_provider->PauseAllTasks();
     }
   }
+
+  // We don't need to do anything further than set the task state to paused
+  // as the next action in an active tool loop will check that state.
+}
+
+void ConversationHandler::ResumeTask() {
+  // If not paused, nothing to do
+  if (tool_use_task_state_ != mojom::TaskState::kPaused) {
+    return;
+  }
+
+  // Set state before calling tool providers so we don't enter a loop via
+  // `OnTaskStateChanged` which calls this function again.
+  tool_use_task_state_ = mojom::TaskState::kRunning;
+
+  for (auto& tool_provider : tool_providers_) {
+    tool_provider->ResumeAllTasks();
+  }
+
+  OnToolUseTaskStateChanged();
+
+  // If a tool is currently in progress, don't call
+  // MaybeRespondToNextToolUseRequest now - it will be called when the tool
+  // completes via RespondToToolUseRequest.
+  if (is_tool_use_in_progress_) {
+    return;
+  }
+
+  MaybeRespondToNextToolUseRequest();
+}
+
+void ConversationHandler::StopTask() {
+  if (tool_use_task_state_ == mojom::TaskState::kNone) {
+    return;
+  }
+
+  tool_use_task_state_ = mojom::TaskState::kStopped;
+  OnToolUseTaskStateChanged();
+
+  for (auto& tool_provider : tool_providers_) {
+    tool_provider->StopAllTasks();
+  }
+}
+
+void ConversationHandler::OnTaskStateChanged(ToolProvider* tool_provider) {
+  // A ToolProvider's task state has changed. Propogate this to the
+  // conversation's task state.
+  DVLOG(4) << "OnTaskStateChanged: task_state="
+           << static_cast<int>(tool_use_task_state_) << " "
+           << is_request_in_progress_ << " " << is_tool_use_in_progress_;
+  if (tool_use_task_state_ == mojom::TaskState::kStopped) {
+    return;
+  }
+
+  if (tool_provider->IsPausedByUser()) {
+    // Pausing from a single tool provider pauses the entire task
+    PauseTask();
+  } else {
+    // Resuming from a single tool provider resumes the entire task
+    ResumeTask();
+  }
```

Comment:
It feels slightly weird to me that tools can be paused/resumed which pauses/resumes other tools. I wonder if it'd be better to have a `OnRequestPause` and `OnRequestResume` and let the conversation be entirely in charge of pausing/resuming.

I take it this is coming from the upstream actor framework